### PR TITLE
chore(broker): move compaction responsibility from LogStream to broker

### LINKED ISSUE #3991: Remove position from snapshot metadata
**Description**

We currently support two modes of replicating snapshots - once after taking it, where the leader pushes the new snapshot out to its followers, and another time when a Raft follower needs to catch up to the leader, at which point the leader sends it the snapshot.

We currently store the snapshot's last processed position as part of the snapshot meta data (that is, as part of the folder name). One of the downsides of this is that on the follower side, we have to open the snapshot in order to figure out the snapshot's last processed position, since Atomix knows nothing about positions and will create snapshot based on the index, term, and timestamp associated with it.

One thing to note here is that we don't actually need the position for anything - we were currently using it as our upper compaction bound, which meant reading it off the snapshot, then mapping that position to an index, and compacting from there. What we can do is just use the index in the snapshot as our upper compaction bound. There is really no advantage to having a delete operation on the log stream abstraction, as compacting is a storage problem, not a stream problem.

In that sense, we should remove the position from the snapshot meta data - after all, we already know the position of the pending snapshot in the `AsyncSnapshotDirector` when we take it and can wait for that to be committed, without encoding it in the snapshot. For compaction, we can simply use the index as our compaction bound and pass this directly on to our Atomix storage.

Not only is this faster for compaction (no more read is necessary), but it alleviates the need to open the snapshot on reception, which has the overhead of starting RocksDB, some threads, allocating memory, triggering possible compaction, etc.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

This PR removes the compaction responsibility from the log stream abstraction, and instead moves it to the broker as a `LogCompactor` API, with currently a single implementation for the Atomix Raft Log, `AtomixLogCompactor`. It also replaces the position from the snapshot interface with a `compactionBound`, which is to be interpreted by the implementation as it sees fit. In the case of this one, it is the upper bound index, such that it is not necessary anymore to open snapshots except for recovery purposes.

Snapshot positions are then only used by the `AsyncSnapshotDirector` for the purpose of committing snapshot, but then we always work off of the positions that are stored in the snapshot itself.

As such, compaction now never has issues with "not found" positions, as the index is just an upper bound, whether or not it is present is irrelevant. Note however that Atomix still guarantees that we always have one segment present, so it's not possible to entirely compact the whole log - as such I think the solution is as safe that our current one.

## Related issues

closes #3991

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by Zelldon** [comment]
> As such, compaction now never has issues with "not found" positions

This should also not happen with the new Index

**Comment by npepinpe** [comment]
Hm, true. I think it's still worth doing but open to challenge.

**Comment by npepinpe** [comment]
Some changes to note which we should discuss:

1. This removes the last processed position from the meta data. This position was not really used anywhere but for logging (e.g. `start taking snapshot from position X`), as the real last processed position was really taken from reading the actual snapshot's stored value for it.
1. The `SnapshotStorage#commitSnapshot(Snapshot)` method was previously return a boolean, but this was never checked - it now returns an `Optional<Snapshot>`, such that the method takes in a pending snapshot and returns an optionally committed snapshot (with a different path). One downside is that a lot of the upstream consumers of this API were actually always expecting things to explode if not committed (since no one was bothering to check the return value), which is meh - so at the moment at some point we short circuit the optional to throw (using `orElseThrow()`). This is not great, but is maintaining the status quo. We can discuss potential options.
1. Snapshots are now sorted/compared by `index` and then `timestamp` - this implies that for the same index, a snapshot taken later is the more "recent" snapshot. This should generally be fine, but ofc is subject to time skew issues - though I guess a lot of our code is so this is probably the least of our worries here :man_shrugging: 
1. The new snapshot name format should be backwards compatible with the previous one - we're a little less strict now and only parse the first 3 parts of the name, ignoring any other parts. Wasn't sure how worth it it is to write a test for this, but we can definitely do so. Though I expect the upgrade tests to catch this kind of stuff.
1. Instead of having everywhere know about "position" or "index", I opted for a more generic "compactionBound" approach, which is interpreted by the snapshot store implementation and the log compactor implementation. This introduces a certain level of coupling between these two, but in a sense as they both operate on the same storage I guess that's inevitable?
1. I had to update tests as some of the interfaces were changed, but I think I kept the same level of integrity/testability - but it would be nice to have someone look over and see if I missed anything :)

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+            new SnapshotMetrics(PARTITION_ID));
+    compactor = new Compactor();
+
+    deletionService = new LogDeletionService(0, PARTITION_ID, compactor, snapshotStorage);
+    actorScheduler.submitActor(deletionService).join();
+  }
+
+  @After
+  public void tearDown() {
+    if (storageReader != null) {
```
Comment (Zelldon) [inline broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: Why do you think they are null?

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+    final var reader = logStorageRule.getRaftLog().openReader(-1);
+
+    // when
+    logStorageRule.appendEntry(1, 1, DATA).index();
+    logStorageRule.appendEntry(2, 2, DATA).index();
+    logStorageRule.appendEntry(3, 3, DATA).index();
+    createSnapshot(5L);
+
+    // then - expect exactly one segment left with entry 3
+    compactor.awaitCompaction(5L, Duration.ofSeconds(5));
```
Comment (Zelldon) [inline broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: Not sure whether this makes sense or not. This would mean if we get a snapshot with index 5 we keep 3, so actually we have an gap in between. Hopefully we will start not  at the beginning :sweat_smile:

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java [hunk 3]
```diff
@@ -78,17 +78,75 @@ protected void onActorStarting() {
...
+    actor.runOnCompletionBlockingCurrentPhase(
+        streamProcessor.getLastProcessedPositionAsync(),
+        (position, error) -> {
+          if (error == null) {
+            lastValidSnapshotPosition = position;
+            LOG.debug(
+                "The position of the last valid snapshot is '{}'. Taking snapshots beyond this position.",
+                lastValidSnapshotPosition);
+          } else {
+            LOG.error(ERROR_MSG_ON_RESOLVE_PROCESSED_POS, error);
```
Comment (Zelldon) [inline engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: Is it correct to continue from here?

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java [hunk 4]
```diff
@@ -78,17 +78,75 @@ protected void onActorStarting() {
...
+        });
   }
 
   @Override
   protected void onActorCloseRequested() {
     logStream.removeOnCommitPositionUpdatedCondition(commitCondition);
   }
 
+  @Override
+  public ActorFuture<Void> closeAsync() {
```
Comment (Zelldon) [inline engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: We have to check our format settings again. In the latest reviews I often saw methods rearrangements. It is is really hard to follow.

File: engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java [hunk 5]
```diff
@@ -575,7 +572,7 @@ public void onRecovered(final ReadonlyProcessingContext context) {
   }
 
   @Test
-  public void shouldNotCreateSnapshotsIfNoProcessorProcessEvent() throws Exception {
+  public void shouldNotCreateSnapshotsIfNoProcessorProcessEvent() {
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: When we could get rid of this limitation then we would also get updates from exporter position.

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java [hunk 6]
```diff
@@ -335,13 +324,13 @@ private int determineInitialPartitionId() {
       final long lastPosition = logReader.seekToEnd();
 
       // dispatcher needs to generate positions greater than the last position
-      int partitionId = 0;
+      int dispatcherPartitionId = 0;
```
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java]: This could cause merge conflicts - I think I renamed it already :D

File: logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java [hunk 7]
```diff
@@ -119,7 +115,7 @@ public long recover() throws Exception {
         // open database to verify that the snapshot is recoverable
         openDb();
         LOG.debug("Recovered state from snapshot '{}'", snapshot);
-        lowerBoundSnapshotPosition = snapshot.getPosition();
+        break;
```
Comment (Zelldon) [inline logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: :see_no_evil: This is often really hard to spot

File: logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java [hunk 8]
```diff
@@ -73,14 +74,14 @@ public void shouldTakeSnapshot() throws Exception {
   }
 
   @Test
-  public void shouldOpenNewDatabaseIfNoSnapshotsToRecoverFrom() throws Exception {
+  public void shouldDoNothingIfNoSnapshotsToRecoverFrom() throws Exception {
     // given
 
     // when
-    final long lowerBoundSnapshotPosition = snapshotController.recover();
+    snapshotController.recover();
```
Comment (Zelldon) [inline logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java]: Should we fail then? I'm wondering what we should do in the processor actually, if we have no snapshot to open (because they are corrupted)

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SnapshotReplicationTest.java [hunk 9]
```diff
@@ -28,7 +29,7 @@
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
-import java.util.zip.CRC32;
+import java.util.zip.CRC32C;
```
Comment (Zelldon) [inline qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SnapshotReplicationTest.java]: :laughing: :+1:

Comment (Zelldon) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SnapshotReplicationTest.java]: We can also use that in the replication chunks right?

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
I think it is good that we were able to remove it on the snapshots etc, but I have a mixed feelings on removing it from the log stream. So the API now makes it only possible to append  and read the log, which is interesting I would say :sweat_smile: Feels like it miss the deletion somehow :man_shrugging: 

That we only need the index to delete is clear to me make I think also makes sense how it works, but this is more an implementation detail I would say. @saig0 any opinion on that?


I added some other comments on the PR.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 1 above)
Comment (npepinpe) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: Force of habit, in case something went wrong before it was initialized and the `@After` still runs - just defensive programming in general.

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java (hunk 3 above)
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: The worse that can happen is that we might take a snapshot when we don't need to, since we fallback to using `-1` as the last valid snapshot position. So if we had a snapshot at position 5, and we process nothing, we will have two snapshots at position 5 (but one with a higher timestamp so they can still be distinguished). It's useless work but I think it's not incorrect. We could also fail, but I'm not sure how is the recovery at the moment, so this seemed like a good compromise.

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java (hunk 4 above)
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: Yes, I reverted those, not sure why it rearranged them :s

File: engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java (hunk 5 above)
Comment (npepinpe) [reply engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: What limitation are you referring to?

File: logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java (hunk 6 above)
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/impl/log/LogStreamImpl.java]: :see_no_evil:

File: logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java (hunk 7 above)
Comment (npepinpe) [reply logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: Hm, the other option is to use a boolean flag, I guess - I'm usually fine with a break but we can use that instead.

File: logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java (hunk 8 above)
Comment (npepinpe) [reply logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java]: This is probably a different issue - current behavior is not to fail. If we have no snapshots at all, then :heavy_check_mark:  If we had some corrupted snapshots, then probably we should step down and throw away our state, but that's a bigger issue, I guess

File: qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SnapshotReplicationTest.java (hunk 9 above)
Comment (npepinpe) [reply qa/integration-tests/src/test/java/io/zeebe/broker/it/clustering/SnapshotReplicationTest.java]: Yes - and also do a checksum of the full snapshot. I'll write an issue for it.

File: broker/src/main/java/io/zeebe/broker/logstreams/AtomixLogCompactor.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+   * @return a future which is completed after compaction is finished
+   */
+  @Override
+  public CompletableFuture<Void> compactLog(final long compactionBound) {
+    Loggers.DELETION_SERVICE.debug("Compacting Atomix log up to index {}", compactionBound);
+    partitionServer.setCompactableIndex(compactionBound);
+    return partitionServer.snapshot().exceptionally(error -> onCompaction(compactionBound, error));
+  }
+
+  private Void onCompaction(final long compactionBound, final Throwable error) {
```
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker/src/main/java/io/zeebe/broker/logstreams/AtomixLogCompactor.java]: > As you wrote in the comment I think it make sense to complete the future exceptionally?

I wasn't sure here, since what exactly would we do with it? Right now nothing, except maybe log it, which we already do here. Though it might be more intuitive if here we return a failed future, and the consumer can decide to log it or not, and we don't log it here, I guess.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 2 above)
Comment (npepinpe) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: This is already the case, though - I just replicated current feature parity :sweat_smile: This is more a side effect of the journal design in Atomix how we use it - there is always one segment left, so even if you try to compact because you received a snapshot ahead of your log, you'll always have some data left over =/

**Comment by npepinpe** [comment]
> So the API now makes it only possible to append and read the log, which is interesting I would say :sweat_smile: Feels like it miss the deletion somehow :man_shrugging:

It really depends on the goal of our `LogStream` abstraction. I realize it was initially meant to be a both an append-only log as well as our main replicated storage layer, but as we transition to using Atomix, it makes sense to me to refocus it as purely an append-only log over an abstracted storage layer (e.g. Atomix). In that sense, compaction and such belongs to the storage layer, I would think - but we can gladly discuss this.

File: broker/src/main/java/io/zeebe/broker/logstreams/AtomixLogCompactor.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (Zelldon) [reply broker/src/main/java/io/zeebe/broker/logstreams/AtomixLogCompactor.java]: > Though it might be more intuitive if here we return a failed future, and the consumer can decide to log it or not, and we don't log it here, I guess.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 1 above)
Comment (Zelldon) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: Hm not sure if this make sense here, it is just noise on reading the test.

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 2 above)
Comment (Zelldon) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: but with truncate we delete it right?

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java (hunk 3 above)
Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: So would could also just skip this and determine the position on the first next snapshot we take?

File: engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java (hunk 5 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: That we do not take snapshots if the position has not changed. I mean we could also just do snapshot every 15min regardless of the position

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 2 above)
Comment (npepinpe) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: I'll double check, but afaik in Atomix, on receiving a snapshot, we reset the head to that index (meaning we truncate down to that index), and start appending. So if we had 1, 2, 3, and we receive snapshot 25, then we truncate nothing and the log will be 1,2,3,26,27, etc.

File: logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java (hunk 7 above)
Comment (Zelldon) [reply logstreams/src/main/java/io/zeebe/logstreams/state/StateSnapshotController.java]: Leave it for now - we will probably rewrite this method in the next issues, because we remove multiple snapshots.

File: logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java (hunk 8 above)
Comment (Zelldon) [reply logstreams/src/test/java/io/zeebe/logstreams/state/StateSnapshotControllerTest.java]: Had a look in the code - if the snapshot was corrupt and the last one we throw an exception, where the processor goes into failed state then. so should be fine

File: engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java (hunk 3 above)
Comment (npepinpe) [reply engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: How would you determine this later? Once we want to take a snapshot, how do we distinguish whether the stream processor has processed events or not?

Comment (Zelldon) [reply engine/src/main/java/io/zeebe/engine/processor/AsyncSnapshotDirector.java]: Maybe we dont need to?

File: engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java (hunk 5 above)
Comment (npepinpe) [reply engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: This is kind of expensive right now because we push snapshots, so with large states this is a lot of extra work for little gain.

That said, it's really tempting to remove it - then we dont have the issue with the exporters. Maybe we should discuss this a bit more thoroughly, it's a little out of scope here imho but a very good point

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: Yes we can do it separately. Came just in my mind when I saw this code :sweat_smile:

Comment (npepinpe) [reply engine/src/test/java/io/zeebe/engine/processor/StreamProcessorTest.java]: https://github.com/zeebe-io/zeebe/issues/4017

File: broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java (hunk 2 above)
Comment (Zelldon) [reply broker/src/test/java/io/zeebe/broker/logstreams/AtomixLogDeletionServiceTest.java]: Hm so we can't start from a clean state

**Review by ChrisKujawa — APPROVED** [review]
